src/cpp/rtps/messages/MessageReceiver.cpp

```diff
@@ -402,124 +411,137 @@ void MessageReceiver::processCDRMsg(
         valid = true;
         uint32_t next_msg_pos = submessage->pos;
         next_msg_pos += (submsgh.submessageLength + 3u) & ~3u;
-        switch (submsgh.submessageId)
+
+        // We ignore submessage if the source participant is to be ignored, unless the submessage king is INFO_SRC
+        // which triggers a reevaluation of the flag.
+        ignore_current_submessage = ignore_submessages &&
+                submsgh.submessageId != INFO_SRC;
+
+        if (!ignore_current_submessage)
         {
-            case DATA:
-            {
-                if (dest_guid_prefix_ != participantGuidPrefix)
-                {
-                    EPROSIMA_LOG_INFO(RTPS_MSG_IN, IDSTRING "Data Submsg ignored, DST is another RTPSParticipant");
-                }
-                else
-                {
-                    EPROSIMA_LOG_INFO(RTPS_MSG_IN, IDSTRING "Data Submsg received, processing.");
-                    valid = proc_Submsg_Data(submessage, &submsgh);
-                }
-                break;
-            }
-            case DATA_FRAG:
-                if (dest_guid_prefix_ != participantGuidPrefix)
-                {
-                    EPROSIMA_LOG_INFO(RTPS_MSG_IN, IDSTRING "DataFrag Submsg ignored, DST is another RTPSParticipant");
-                }
-                else
-                {
-                    EPROSIMA_LOG_INFO(RTPS_MSG_IN, IDSTRING "DataFrag Submsg received, processing.");
-                    valid = proc_Submsg_DataFrag(submessage, &submsgh);
-                }
-                break;
-            case GAP:
-            {
-                if (dest_guid_prefix_ != participantGuidPrefix)
-                {
-                    EPROSIMA_LOG_INFO(RTPS_MSG_IN, IDSTRING "Gap Submsg ignored, DST is another RTPSParticipant...");
-                }
-                else
-                {
-                    EPROSIMA_LOG_INFO(RTPS_MSG_IN, IDSTRING "Gap Submsg received, processing...");
-                    valid = proc_Submsg_Gap(submessage, &submsgh);
-                }
-                break;
-            }
-            case ACKNACK:
+
+            switch (submsgh.submessageId)
             {
-                if (dest_guid_prefix_ != participantGuidPrefix)
+                case DATA:
```

Comment:
Processing a DATA submessage for the PDP builtin endpoint may result in adding the current `source_guid_prefix_` to the list of ignored participants.

This means we should reevaluate `ignore_submessages` in that case